pkg/concurrency/stoppable.go

```diff
@@ -0,0 +1,82 @@
+package concurrency
+
+// NewStoppable creates a new Stoppable for arranging a graceful shutdown.
+func NewStoppable() Stoppable {
+	return &stoppableImpl{
+		stop:    NewErrorSignal(),
+		stopped: NewErrorSignal(),
+	}
+}
+
+// Stoppable encapsulates stop and stopped signals for arranging a graceful shutdown sequence for goroutines (and async
+// processing in general).
+type Stoppable interface {
+	// Client returns an interface that should be used outside the goroutine to request a graceful shutdown and check
+	// the shutdown status.
+	Client() StoppableClient
+
+	// Flow returns an interface that should be used in goroutine control flow to check and modify the state of
+	// stoppable. Flow should not be used outside the goroutine (or async processing).
+	Flow() StoppableFlow
+}
+
+// StoppableClient represents an interface for a client external to goroutine (or async process) to request it to
+// gracefully shutdown and to check the shutdown status.
+type StoppableClient interface {
+	// Stop requests stoppable to shut down (or stop).
+	Stop()
+
+	// Stopped returns ReadOnlyErrorSignal that can be used for waiting until the stoppable reports that it finally
+	// stopped.
+	Stopped() ReadOnlyErrorSignal
+}
+
+// StoppableFlow represents an interface for goroutine to check and report status of a graceful shutdown.
+type StoppableFlow interface {
+	// StopWithError initiates a shutdown due to an error that happened during async processing in goroutine.
+	StopWithError(error)
```

Comment:
Shouldn't `StoppableClient.Stop` and `StoppableFlow.StopWithError` be closer together in the same object?

Reply:
This was intentional.

`.Client().Stop()` is called externally (outside of goroutine), and I haven't found any cases when non-`nil` error was provided as parameter. That's why `.Client().Stop()` doesn't take `error` parameters.

OTOH, I found goroutines self-stop due to error (e.g. a channel closed). That's why I added `.Flow().StopWithError(error)` for internal interface of goroutines (that's what I call `StoppableFlow`).